### PR TITLE
[Behat][Taxon] Adding taxon to specific parent

### PR DESCRIPTION
**Pull request by Zales0123**
| Q               | A
| --------------- | ---
| Bug fix?        | no
| New feature?    | yes
| BC breaks?      | yes
| Related tickets | |
| License         | MIT

zrzut ekranu 2016-11-04 o 14 33 53

### CODE REVIEW

File: features/taxonomy/managing_taxons/adding_taxon.feature [hunk 1]
```diff
@@ -27,14 +27,4 @@ Feature: Adding a new taxon
...
-    Scenario: Adding a new taxon with parent taxon
-        Given the store has "Category" taxonomy
-        And I want to create a new taxon
-        When I specify its code as "stickers"
-        And I name it "Stickers" in "English (United States)"
-        And I choose "Category" as a parent taxon
-        And I add it
-        Then I should be notified that it has been successfully created
-        And the "Stickers" taxon should appear in the registry
+        And it should not be belong to any other taxon
```
Comment (TheMadeleine) [inline features/taxonomy/managing_taxons/adding_taxon.feature]: ``And it should not belong to any other taxon`` 🎉

File: src/Sylius/Behat/Context/Ui/Admin/ManagingTaxonsContext.php [hunk 2]
```diff
@@ -231,6 +231,19 @@ public function thisTaxonShouldBelongsTo(TaxonInterface $taxon)
     }
 
     /**
+     * @Given it should not be belong to any other taxon
+     */
+    public function itShouldNotBeBelongToAnyOtherTaxon()
```
Comment (TheMadeleine) [inline src/Sylius/Behat/Context/Ui/Admin/ManagingTaxonsContext.php]: ``itShouldNotBelongToAnyOtherTaxon()``

File: src/Sylius/Bundle/AdminBundle/Resources/views/Taxon/Partial/_treeWithButtons.html.twig [hunk 3]
```diff
@@ -5,7 +5,7 @@
     {% import '@SyliusUi/Macro/buttons.html.twig' as buttons %}
     {% import _self as tree %}
 
-    {% for taxon in taxons %}
+    {% for taxon in taxons if taxon.id != null %}
```
Comment (pamil) [inline src/Sylius/Bundle/AdminBundle/Resources/views/Taxon/Partial/_treeWithButtons.html.twig]: Why is it needed?

File: src/Sylius/Component/Taxonomy/Model/Taxon.php [hunk 4]
```diff
@@ -176,10 +179,12 @@ public function hasChild(TaxonInterface $taxon)
     public function addChild(TaxonInterface $taxon)
     {
         if (!$this->hasChild($taxon)) {
-            $taxon->setParent($this);
-
             $this->children->add($taxon);
         }
+
+        if ($this !== $taxon->getParent()) {
```
Comment (pamil) [inline src/Sylius/Component/Taxonomy/Model/Taxon.php]: This behaviour is worth specifying :)

File: src/Sylius/Behat/Page/Admin/Taxon/CreateForParentPage.php [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+class CreateForParentPage extends CreatePage implements CreateForParentPageInterface
+{
+    /**
+     * {@inheritdoc}
+     */
+    public function getRouteName()
+    {
+        return 'sylius_admin_taxon_create_for_parent';
+    }
+
```
Comment (pamil) [inline src/Sylius/Behat/Page/Admin/Taxon/CreateForParentPage.php]: Unnecessary blank line.

File: src/Sylius/Bundle/AdminBundle/Resources/views/Taxon/_form.html.twig [hunk 6]
```diff
@@ -2,7 +2,11 @@
     {{ form_errors(form) }}
     <div class="two fields">
         {{ form_row(form.code) }}
-        {{ form_row(form.parent) }}
+        {% if taxon.id != null %}
+            {{ form_row(form.parent) }}
+        {% else %}
+            {{ form_widget(form.parent, {'attr': {'style': 'display: none;'}}) }}
```
Comment (pjedrzejewski) [inline src/Sylius/Bundle/AdminBundle/Resources/views/Taxon/_form.html.twig]: Proper fix would be to remove this field completely. We do not use it anymore, you should always add child node via button.

Comment (Zales0123) [inline src/Sylius/Bundle/AdminBundle/Resources/views/Taxon/_form.html.twig]: Yup, but what if we want to change parent of existing taxon? Or is it the feature we don't want to support?

Comment (pjedrzejewski) [inline src/Sylius/Bundle/AdminBundle/Resources/views/Taxon/_form.html.twig]: Good point... :D So let's leave this for now... And we will iterate on this.

**Pull request merged**